quill-async/src/main/scala/io/getquill/sources/async/AsyncSource.scala

```diff
@@ -96,4 +97,8 @@ abstract class AsyncSource[D <: SqlIdiom, N <: NamingStrategy, C <: Connection](
       }
     }
   }
+
+  def querySingle[T](sql: String, bind: BindedStatementBuilder[List[Any]] => BindedStatementBuilder[List[Any]], extractor: RowData => T)(implicit ec: ExecutionContext) = {
+    query(sql, bind, extractor).map(_.head)
```

Comment:
Let's give a better error message using:

Source.scala
```scala
def handleSingleResult[T](list: List[T]) =
  l match {
    case value :: Nil => value
    case other => throw new IllegalStateException(s"Expected a single result but got $other")
  }
```
AsyncSource.scala and others

```scala
query(sql, bind, extractor).map(handleSingleResult)
```

You could share the implementation adding a method with this implementation to `Source`.